src/beatree/ops/update.rs

```diff
@@ -0,0 +1,196 @@
+    }
+}
+
+struct BranchChanges {
+    prev: Option<PrevBranch>,
+    changes: Vec<BranchChange>,
+}
+
+impl BranchChanges {
+    fn is_empty(&self) -> bool {
+        self.changes.is_empty()
+    }
+
+    // Apply the changes. Returns `Some` if there are leftover changes to be applied
+    // to the first branch following the returned key. 
+    fn apply(
+        self, 
+        bbn_index: &mut Index,
+        bnp: &mut branch::BranchNodePool,
+        bbn_store_writer: &mut bbn::BbnStoreWriter,
+    ) -> Option<(Key, BranchChanges)> {
+        if self.is_empty() { return None }
+
+        let existing_keys = self.prev.as_ref().into_iter().flat_map(|p| {
+            let prefix = p.node.prefix();
+            (0..p.node.n()).map(|i| p.node.separator(i as usize)).map(move |separator| reconstruct_key(prefix, separator))
+        });
+
+        let merge = existing_keys.merge_join_by(
+            self.changes.iter(),
+            |existing_key, change| existing_key.cmp(change.key())
+        );
+
+        let mut gauge = BranchGauge::new();
+
+        for merge_item in merge.clone() {
+            match merge_item {
+                EitherOrBoth::Left(key) => {
+                    gauge.ingest(key);
+                }
+                EitherOrBoth::Right(BranchChange::Insert(key, _)) => {
+                    gauge.ingest(*key);
+                }
+                EitherOrBoth::Both(key, BranchChange::Remove(_)) => {}
+                EitherOrBoth::Both(key, BranchChange::Update(_, _)) => {
+                    gauge.ingest(key);
+                }
+                _ => panic!("invalid branch update"),
+            }
+        }
+
+        let new_size = gauge.body_size();
+
+        if new_size > BRANCH_NODE_BODY_SIZE {
+            // TODO: split
+            todo!()
+        } else if new_size <= BRANCH_MERGE_THRESHOLD {
+            // TODO: collect all existing keys/updates/inserts into branch changes
+            // and pass to next branch
+            todo!()
+        } else {
+            // TODO: just create a replacement branch
+            todo!()
+        }
+
+        // TODO: delete previous branch.
+    }
+}
+
+struct BranchGauge {
+    first_separator: Option<Key>,
+    prefix_len: usize,
+    // total separator len, not counting prefix.
+    separator_len: usize,
+    n: usize,
+}
+
+impl BranchGauge {
+    fn new() -> Self {
+        BranchGauge {
+            first_separator: None,
+            prefix_len: 0,
+            separator_len: 0,
+            n: 0,
+        }
+    }
+
+    fn ingest(&mut self, key: Key) {
+        let Some(ref first) = self.first_separator else {
+            self.first_separator = Some(key);
+            self.separator_len = separator_len(&key);
+            self.prefix_len = self.separator_len;
+
+            self.n = 1;
+            return
+        };
+
+        self.separator_len = std::cmp::max(separator_len(&key), self.separator_len);
+        self.prefix_len = prefix_len(first, &key);
```

Comment:
Initially, I thought this was wrong because perhaps the ingestion of a new key could make the `prefix_len` smaller, leading to some information loss. However, this is not possible if the keys are ingested in order. Therefore, adding a comment stating this requirement could be helpful for the future

Reply:
The prefix len can get smaller even though keys are being ingested in lexicographical order, by my understanding. What's more important is that the `separator_len` includes the prefix length, in this struct.

If you have 3 keys `0xAA, 0xAB, 0xBA`, the first two keys have prefix `0xA` and adding the third key reduces the prefix by a bit.